### PR TITLE
Docker build context

### PR DESCRIPTION
**Pull request by agnihotrigaurav**
Jira ID WRKR-386

Problem Description.
docker-build forces docker build context to project root.

Problem solution
Added new property context in docker argument list to specify build context for docker-build.
Currently all the file below source/ are added to tar file used for creating docker-build. 

Change would take into consideration value of context field as well. It will add files in directory "source/"+context

Testing
Automated testcase added for same.

Staging testing (Repository - https://github.com/agnihotrigaurav/test-docker-build)

01) Test pass when context is correctly specified.Result(Pass)
URL - https://dev.wercker.com/gauagnih/test-docker-build/runs/build/5bf3e41cef658a00082dfd9a
-------------------------------------------------------------------------------------------
02) In case context path does not exist. Test case fails with message(Error response from daemon: Cannot locate specified Dockerfile: Dockerfile)
URL - https://dev.wercker.com/gauagnih/test-docker-build/runs/build/5bfcdf0ac830fd0006fda8a2?step=5bfcdf352ce5e50006c28c96
-------------------------------------------------------------------------------------------
03) Extra forward slash in context path. Result(Pass)
URL - https://dev.wercker.com/gauagnih/test-docker-build/runs/build/5bfd21c76b89c50006c67cfe
-------------------------------------------------------------------------------------------
04) Regression case when context is not specified. Test case pass as all the files are present in current folder.Result(Pass)
URL - https://dev.wercker.com/gauagnih/wercker-docker-push/runs/build/5bfd210803623a00074eecdf

### CODE REVIEW

File: docker/docker_build.go [hunk 1]
```diff
@@ -372,11 +385,9 @@ func (s *DockerBuildStep) buildInputTar(sourceTar string, destTar string) error
 		}
 
 		// copy files from /source into the root of the new tar
```
Comment (nigeldeakin) [inline docker/docker_build.go]: Copy files under the specified build context into the root of the new tar

File: docker/docker_build.go [hunk 2]
```diff
@@ -351,6 +359,11 @@ func (s *DockerBuildStep) buildInputTar(sourceTar string, destTar string) error
 	tr := tar.NewReader(artifactReader)
 	tw := tar.NewWriter(layerFile)
 
+	dockerBuildContext := "source/"
+	if s.dockerBuildContext != "" {
+		dockerBuildContext = fmt.Sprintf("%s%s/", dockerBuildContext, s.dockerBuildContext)
```
Comment (nigeldeakin) [inline docker/docker_build.go]: Please use `filepath.Join`

File: docker/docker_build.go [hunk 3]
```diff
@@ -37,19 +37,20 @@ import (
 // DockerBuildStep needs to implemenet Step
```
[the remark this answers is not in the document]
Comment (agnihotrigaurav) [reply docker/docker_build.go]: Done...

File: docker/docker_build.go [hunk 4]
```diff
@@ -175,6 +176,13 @@ func (s *DockerBuildStep) configure(env *util.Environment) error {
 		}
 		s.authConfigs = authConfigs
 	}
+
+	if dockerBuildContext, ok := s.data["context"]; ok {
+		s.dockerBuildContext = env.Interpolate(dockerBuildContext)
+		if last := len(s.dockerBuildContext) - 1; last >= 0 && s.dockerBuildContext[last] == '/' {
+			s.dockerBuildContext = s.dockerBuildContext[:last]
+		}
```
[the remark this answers is not in the document]
Comment (agnihotrigaurav) [reply docker/docker_build.go]: "filepath.Join" already takes care of slash....just removing code for manipulating same.

File: docker/docker_build.go (hunk 1 above)
Comment (agnihotrigaurav) [reply docker/docker_build.go]: Done.

File: docker/docker_build.go (hunk 2 above)
Comment (agnihotrigaurav) [reply docker/docker_build.go]: Done.

**Comment by agnihotrigaurav** [comment]
As discussed I have created another jira to address base-path issue. Current PR only address issue related to context. If docker build context is different from  $WERCKER_ROOT, then context is updated correctly.

File: changelog.md [hunk 5]
```diff
@@ -1,5 +1,7 @@
 ## unreleased
 
+- Fix issue with docker build context (#537)
```
Comment (nigeldeakin) [inline changelog.md]: I suggest 
Add new property "context" to internal/docker-build
(This is a new feature rather than a fix)

Comment (nigeldeakin) [reply changelog.md]: This isn't important, so please do what you think best. 

A "bug" is when the product does not behave as described in the documentation. 

The current docs state
> The build’s context is the set of files in the $WERCKER_ROOT directory of the currently-running pipeline container.
This is exactly what the product does now.

This PR is extending the functionality of this step to allow the context to be configured. That is an excellent improvement, but the current behaviour is deliberate, not a bug.

Comment (nigeldeakin) [reply changelog.md]: (You will, of course, need to update that sentence)

Comment (agnihotrigaurav) [reply changelog.md]: Done.

**Pull request merged**